news/_posts/2019-15-01-resilient-php-applications.md

```diff
@@ -0,0 +1,108 @@
+
+But, no matter how well designed is your project and how many tests you have, at some point your website can become unavailable.
+
+This can be because of your database, or calls to third party services (payment methods API, PrestaShop REST API to retrieve modules information) are unavailable.
+
+Instead of thinking everything will be ok, we need - as developers - to manage the failures of our systems.
+
+We need to create resilient systems, to create websites able to recover quickly from network disasters.
+
+## The risks of depending to external services
+
+Assuming all external services will works fine is risky because you don't think
+about what should happens in case of unavailability:
+
+* if the payment APIs are not reachable, instead of freezing your shop you could suggest your customer to contact them back once the system is ok and provide them a little discount;
+
+* you should be able to install PrestaShop even if the translations APIs are not reachable,
+at least in english;
+
+* you should be able to access your back office, no matter how many external APIs are not available;
+
+**The collapse of a third party service shouldn't collapse your shop.**
+
+Another issue with this kind of naive assumptions is that your system will try to contact again and again the third party service. This is slowing your system and make the user experience terrible when you could manage to provide an alternative response.
+
+> There is no good reason to provide terrible user experience! period.
+
+## Introducing the PrestaShop Circuit Breaker library
+
+The [Circuit Breaker](https://martinfowler.com/bliki/CircuitBreaker.html) design pattern provides strategies and functions to manage the unavailability of the third party services.
+
+The basic idea behind the circuit breaker is very simple:
+
+* every call of third party service is wrapped into a function call from the Circuit Breaker;
+* once we reach a defined numbers of failures (means the service is unreachable), the Circuit Breaker is "Open": an alternative response is provided and no calls to the service are done;
+* after a defined amount of time (called "reset timeout" sometimes), the Circuit Breaker try to access the service: if the service is available it will be used again, else the alternative response is provided instead;
+
+![https://martinfowler.com/bliki/images/circuitBreaker/sketch.png](https://martinfowler.com/bliki/images/circuitBreaker/sketch.png)
+
+> Schema from Martin Fowler's blog.
+
+This is also a big change of mindset for both the e-merchants and the developers, because for each process they need to think about what should be done in case of unavailability of a service.
+
+Like we already do in some old web browsers for the design of our websites, we should think about degraded experience when some third party services we use are not availables.
```

Comment:
who is "we"?
in general, please try to avoid to use "we" when it's not clear who is behind, and find a different way to write the sentence.
Example: "Like good developers do to handle old web browsers when they design a website, it's possible to think from the beginning about a "degraded experience" when some third party services are not available".

Reply:
I like your suggestion 👍

Reply:
"Like web developers do to handle old web browers when they design a website, it's possible to thing from the beginning about a "degraded experience" when some third party services are not available, using a "circuit breaker"."